src/solvers/equation_solver.hpp

```diff
@@ -49,52 +49,93 @@ class EquationSolver : public mfem::Solver {
    * @note Implements mfem::Operator::SetOperator
    */
-  void SetOperator(const mfem::Operator& op) override
-  {
-    solver().SetOperator(op);
-    height = solver().Height();
-    width  = solver().Width();
-  }
+  void SetOperator(const mfem::Operator& op) override;
+
+  /**
+   * @brief An overload for "intercepting" HypreParMatrices
+   * such that they can be converted to a SuperLURowLocMatrix
+   * when running in SuperLU mode
+   * @param[in] op The operator (system matrix) to use, "A" in Ax = b
+   */
+  void SetOperator(const mfem::HypreParMatrix& matrix);
 
   /**
    * Solves the system
    * @param[in] b RHS of the system of equations
    * @param[out] x Solution to the system of equations
    * @note Implements mfem::Operator::Mult
    */
-  void Mult(const mfem::Vector& b, mfem::Vector& x) const override { solver().Mult(b, x); }
+  void Mult(const mfem::Vector& b, mfem::Vector& x) const override;
 
   /**
    * Returns the underlying solver object
-   * @return The underlying nonlinear solver, if one was configured
-   * when the object was constructed, otherwise, the underlying linear solver
+   * @return A non-owning reference to the underlying nonlinear solver
    */
-  mfem::IterativeSolver&       solver() { return (nonlin_solver_) ? **nonlin_solver_ : *iter_lin_solver_; }
-  const mfem::IterativeSolver& solver() const { return (nonlin_solver_) ? **nonlin_solver_ : *iter_lin_solver_; }
+  mfem::IterativeSolver&       nonlinearSolver() { return *nonlin_solver_; }
+  const mfem::IterativeSolver& nonlinearSolver() const { return *nonlin_solver_; }
 
   /**
-   * Returns the underlying linear solver object, even if the class instance
-   * has been configured as a nonlinear solver
-   * @return The underlying linear solver
+   * Returns the underlying linear solver object
+   * @return A non-owning reference to the underlying linear solver
    */
-  mfem::IterativeSolver&       linearSolver() { return *iter_lin_solver_; }
-  const mfem::IterativeSolver& linearSolver() const { return *iter_lin_solver_; }
+  mfem::Solver& linearSolver()
+  {
+    mfem::Solver* result;
+    std::visit([&result](auto&& solver) { result = solver.get(); }, lin_solver_);
+    return *result;
+  }
+  const mfem::Solver& linearSolver() const
+  {
+    mfem::Solver* result;
+    std::visit([&result](auto&& solver) { result = solver.get(); }, lin_solver_);
+    return *result;
+  }
 
 private:
-  std::unique_ptr<mfem::IterativeSolver>                iter_lin_solver_;
-  std::optional<std::unique_ptr<mfem::IterativeSolver>> nonlin_solver_;
-  std::unique_ptr<mfem::Solver>                         prec_;
+  /**
+   * @brief Builds an iterative solver given a set of linear solver parameters
+   * @param[in] comm The MPI communicator object
+   * @param[in] lin_params The parameters for the linear solver
+   */
+  static std::unique_ptr<mfem::IterativeSolver> buildIterLinSolver(MPI_Comm                      comm,
```

Comment:
Style issue: I generally like verbose names for methods. While this is fairly obvious, I think `buildIterativeLinearSolver` would be a more descriptive name.